### ISSUE TITLE
tf.js uses JavaScript `eval()`, making it impossible to use TensorFlow JS in websites with stricter Content Security Policy

### ISSUE THREAD

**Issue by pbelcak** [tensorflow/tfjs#7554]
Hello,

Several threads on this are already scattered over the internet (e.g. [here](https://stackoverflow.com/questions/56618633/cant-import-tensorflow-js-in-chrome-extension)).

In short, the default implementation of tf.js seems to use `eval()` somewhere, probably for performance reasons. Since `script-src 'unsafe-eval'` is no longer an acceptable content security policy for a Chromium extension, the use of `eval()` effectively makes it impossible to use tf.js in Chromium extension, which is a massive bugger.

Is there any way of using tf.js without it calling `eval()`? Performance degradation would not be an issue.

### Minimal Reproducible Example (for a Chromium extension):

Simply load the folder containing these four files by pressing `Load unpacked` in `Extensions` after enabling the extensions developer mode. Then, click on the extension under your browse bar to activate it and load `index.html`.

tfjs.js
```
downloaded from https://cdn.jsdelivr.net/npm/@tensorflow/tfjs
```

index.html
```
<!DOCTYPE html>
<html>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<script src="tfjs.js"> </script>

<head>
    <title>MRE</title>
    <meta charset="utf-8">
</head>
<body>
</body>
<script type="module" src="script.js"></script>
</html>
```

script.js

```
// Create a simple model.
const model = tf.sequential();
```

manifest.json
```
{
    "name": "MRE",
    "version": "0.0.1",
    "description": "Minimal Reproducible Example",
    "manifest_version": 3,
    "author": "me",
    "action":{
        "default_popup": "index.html",
        "default_title": "MRE"
    },
    "permissions": [
        "activeTab"
    ],
    
    "content_security_policy": {
        "extension_pages": "script-src 'self'"
    }
}
```

Error given by the browser
```
Uncaught EvalError: Refused to evaluate a string as JavaScript because 'unsafe-eval' is not an allowed source of script in the following Content Security Policy directive: "script-src 'self'".
```

**Comment by mattsoulanille** [comment]
Thanks for the issue report! The example code seems to fail for any script I try to load, .e.g.
```html
<script src="https://cdn.jsdelivr.net/npm/is-number@7.0.0/index.min.js"></script>
```

```
Refused to load the script 'https://cdn.jsdelivr.net/npm/is-number@7.0.0/index.min.js' because it violates the following Content Security Policy directive: "script-src 'self' 'wasm-unsafe-eval' 'inline-speculation-rules' http://localhost:* http://127.0.0.1:*". Note that 'script-src-elem' was not explicitly set, so 'script-src' is used as a fallback.
```

Are you sure this is a TFJS issue?

**Comment by Neilblaze** [comment]
@pbelcak Importing `index.min.js` locally instead of using CDN should apparently fix the issue.

**Comment by pbelcak** [comment]
My apologies gentlemen, I oversimplified the MRE.

I have updated the original post. This is the error that occurs the moment you engage tf in your code (regardless if it is stored locally or loaded from a remote server with a more lax CSP.

This is the exact error you click on the extension to activate it.
```
Uncaught EvalError: Refused to evaluate a string as JavaScript because 'unsafe-eval' is not an allowed source of script in the following Content Security Policy directive: "script-src 'self'".
```

@mattsoulanille Please see the updated example in the original post.

@Neilblaze unfortunately this was just a mistake of my doing. I corrected the MRE, the issue is not where the script is loaded from but that `eval`, `new Function`, or something equivalent is being called.

**Comment by mattsoulanille** [comment]
Found the cause! This is due to the use of the `Function` constructor instead of `eval`, which is why I couldn't find it before. It looks like regenerator-runtime constructs a function at tfjs.js:17:159850. 

You can try loading the [es2017 TFJS bundle](https://cdn.jsdelivr.net/npm/@tensorflow/tfjs@4.4.0/dist/tf.es2017.js) that doesn't use regenerator-runtime for polyfills. I've tested this with your example code and confirmed it can create an empty model. I also tried adding two tensors together, and that worked too.

**Comment by gaikwadrahul8** [comment]
Hi, @pbelcak 

Apologize for the delayed response and please refer this https://github.com/tensorflow/tfjs/issues/7554#issuecomment-1506110153 from @mattsoulanille it will resolve your issue and if this issue got resolved please feel free to close this issue ? Thank you!

**Comment by pbelcak** [comment]
Hi @mattsoulanille @gaikwadrahul8,

Thanks, this seems to fix the issue, though I haven't had the chance to test it on more than the MRE. Before we close this, is there any way/place where I could add this to the tfjs docs, especially on the website where all the versions are listed? Before coming here I found at least 3 StackOverflow and forum questions trying to deal with this.

**Comment by Tiag0F** [comment]
> Found the cause! This is due to the use of the `Function` constructor instead of `eval`, which is why I couldn't find it before. It looks like regenerator-runtime constructs a function at tfjs.js:17:159850.
> 
> You can try loading the [es2017 TFJS bundle](https://cdn.jsdelivr.net/npm/@tensorflow/tfjs@4.4.0/dist/tf.es2017.js) that doesn't use regenerator-runtime for polyfills. I've tested this with your example code and confirmed it can create an empty model. I also tried adding two tensors together, and that worked too.

I'm having the same problem and tried to use the [es2017 TFJS bundle](https://cdn.jsdelivr.net/npm/@tensorflow/tfjs@4.4.0/dist/tf.es2017.js) but when I try to load a model using tf.loadLayersModel, it says that tf.loadLayersModel is not a function. I'm importing the bundle like this
import * as tf from "/resources/tf.es2017.js";
inside a background.js (service worker)


Also, do you know if this 
import * as tf from "https://cdn.jsdelivr.net/npm/@tensorflow/tfjs@3.11.0/dist/tf.fesm.js"; 
is similar to the one you used? I'm new to this and very confused about what is going on. Any help would be appreciated.
My goal is to make chrome extension that loads a keras model using tensorflowJS and predict something.

**Comment by mattsoulanille** [comment]
@Tiag0F That looks right to me. Here's a sample of how I got it working on my machine:

index.html
```html
<head>
</head>
<body>
  <script src="./index.js" type="module"></script>
</body>
```

index.js
```js
import * as tf from "https://cdn.jsdelivr.net/npm/@tensorflow/tfjs@3.11.0/dist/tf.fesm.js";

window.tf = tf;

console.log(tf.loadLayersModel);
```

Console prints:
```
ƒ loadLayersModel(pathOrIOHandler, options) {
    if (options == null) {
        options = {};
    }
    return loadLayersModelInternal(pathOrIOHandler, options);
}
```
I can also access the `tf` variable from the console since I assigned `window.tf = tf` (this is just for debugging).

Serving this with `python -m http.server` lets me load the webpage and prints the value of `tf.loadLayersModel`. This might work differently in a chrome addon, and this example assumes you're using esmodules instead of commonjs or a bundle. I also recommend using a newer version of TFJS since 3.11.0 is rather old. 
```js
import * as tf from "https://cdn.jsdelivr.net/npm/@tensorflow/tfjs@4.5.0/dist/tf.fesm.js";
```

**Comment by dhruv53** [comment]
I am facing the same issue. Creating a chrome extension but whenever I am calling any version of tensorflow.js using cdn, I get an error:

> Refused to load the script 'https://cdn.jsdelivr.net/npm/@tensorflow/tfjs@3.11.0/dist/tf.fesm.js' because it violates the following Content Security Policy directive: "script-src 'self' 'wasm-unsafe-eval' 'inline-speculation-rules' http://localhost:* http://127.0.0.1:*". Note that 'script-src-elem' was not explicitly set, so 'script-src' is used as a fallback.

I have even tried the approach suggested by @mattsoulanille, but the issue still exist.

popup.html
```
<!DOCTYPE html>
<html>
<head>
  <title>Webcam Feed</title>
  <style>
    canvas {
      width: 300px;
      height: 200px;
    }
  </style>
</head>
<body>
  <h1>Hello</h1>
  <canvas id="canvas"></canvas>
  <script src="popup.js" type="module"></script>
  <!-- Tensorflow added below -->
  <script src="./test.js" type="module"></script>
</body>
</html>
```
test.js
```
import * as tf from "https://cdn.jsdelivr.net/npm/@tensorflow/tfjs@3.11.0/dist/tf.fesm.js";

window.tf = tf;

console.log(tf.loadLayersModel);
```

**Comment by mattsoulanille** [comment]
For a Chrome extension, I think you may need to download the tfjs bundle and include it in your local files for the extension. Does it work when you do that?

**Comment by pbelcak** [comment]
> I am facing the same issue. Creating a chrome extension but whenever I am calling any version of tensorflow.js using cdn, I get an error:
> 
> > Refused to load the script 'https://cdn.jsdelivr.net/npm/@tensorflow/tfjs@3.11.0/dist/tf.fesm.js' because it violates the following Content Security Policy directive: "script-src 'self' 'wasm-unsafe-eval' 'inline-speculation-rules' http://localhost:* http://127.0.0.1:*". Note that 'script-src-elem' was not explicitly set, so 'script-src' is used as a fallback.
> 
> I have even tried the approach suggested by @mattsoulanille, but the issue still exist.
> 
> popup.html
> 
> ```
> <!DOCTYPE html>
> <html>
> <head>
>   <title>Webcam Feed</title>
>   <style>
>     canvas {
>       width: 300px;
>       height: 200px;
>     }
>   </style>
> </head>
> <body>
>   <h1>Hello</h1>
>   <canvas id="canvas"></canvas>
>   <script src="popup.js" type="module"></script>
>   
>   <script src="./test.js" type="module"></script>
> </body>
> </html>
> ```
> 
> test.js
> 
> ```
> import * as tf from "https://cdn.jsdelivr.net/npm/@tensorflow/tfjs@3.11.0/dist/tf.fesm.js";
> 
> window.tf = tf;
> 
> console.log(tf.loadLayersModel);
> ```

Hi @dhruv53, this seems to be just a violation of the CSP. Your CSP allows self, localhost, and `127.0.0.1` as valid sources, yet you seem to be trying to get the tfjs from a remote CDN endpoint. Either download the tfjs and include it with your application, or write a more lenient CSP.

**Comment by pbelcak** [comment]
**TL;DR** for anyone who stumbles upon this in the future. 

There are several variants of TFJS, some of which use `new Function` to ensure browser backward compatibility. What you want to use in your extension is a variant that does not do these "unsafe" inline function hacks. The es2017 TFJS bundle is one of them, and using it fixes the issue.

https://[cdn.jsdelivr.net/npm/@tensorflow/tfjs@4.4.0/dist/tf.es2017.js](https://cdn.jsdelivr.net/npm/@tensorflow/tfjs@4.4.0/dist/tf.es2017.js)

**Comment by pbelcak** [comment]
@arnoutvreugdenhil That's a good point. Is the error you're getting as a result the same as the one above for browser extensions? Could you provide a minimal example?

**Comment by arnoutvreugdenhil** [comment]
My version of electron is older (20) and seems to still accept unsafe-eval in CSP which of course is pretty unsafe but is ok as a temporary workaround during development.
It would be better to fix it in the repository.

```
Uncaught EvalError: Refused to evaluate a string as JavaScript because 'unsafe-eval' is not an allowed source of script in the following Content Security Policy directive: "script-src 'self' 'unsafe-inline'".

    at Function (<anonymous>)
    at node_modules/@tensorflow/tfjs-node/dist/proto/api_pb.js line 16:14
```

I can't provide a clean reproduction right now but I'm using vite + vue + electron with vladmandic's human lib.
`npm i @tensorflow/tfjs-node @vladmandic/human`
That will require installing aws-sdk, mock-aws-s3 and nock too btw.
No idea if Human has any effect on whether or not api_bp.js is loaded or not.

**Comment by pbelcak** [comment]
It looks like the only solution to this is to provide a version of `tfjs-node` that does not use `new Function`.

Thoughts?

**Comment by arnoutvreugdenhil** [comment]
I guess so...
The contents of the file say it's generated code and it's a few years old, so maybe google can generate something better nowadays.

**Comment by shimaamorsy** [comment]
I am facing the same problem, what should I do to solve this problem?

**Comment by pbelcak** [comment]
> I am facing the same problem, what should I do to solve this problem?

https://github.com/tensorflow/tfjs/issues/7554#issuecomment-1602563405
